Add lowering path for LinearReLU module

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71427

Summary: This commit adds a lowering path for the LinearReLU modules
in static quantization mode. This includes torch.nn.qat.Linear,
torch.nn.intrinsic.LinearReLU, and torch.nn.intrinsic.qat.LinearReLU.
Future commits will add support for dynamic quantization and functional
LinearReLU.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_linear_module

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T107568666

Differential Revision: [D33694742](https://our.internmc.facebook.com/intern/diff/D33694742)

andrewor14 committed:
Add lowering path for LinearReLU module

Summary: This commit adds a lowering path for the LinearReLU modules
in static quantization mode. This includes torch.nn.qat.Linear,
torch.nn.intrinsic.LinearReLU, and torch.nn.intrinsic.qat.LinearReLU.
Future commits will add support for dynamic quantization and functional
LinearReLU.

Test Plan: python test/test_quantization.py TestQuantizeFx.test_linear_relu_lowering

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T107568666

[ghstack-poisoned]

andrewor14 committed:
Update on "Add lowering path for LinearReLU module"


Summary: This commit adds a lowering path for the LinearReLU modules
in static quantization mode. This includes torch.nn.qat.Linear,
torch.nn.intrinsic.LinearReLU, and torch.nn.intrinsic.qat.LinearReLU.
Future commits will add support for dynamic quantization and functional
LinearReLU.

Test Plan: python test/test_quantization.py TestQuantizeFx.test_linear_relu_lowering

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T107568666

[ghstack-poisoned]

andrewor14 committed:
Update on "Add lowering path for LinearReLU module"


Summary: This commit adds a lowering path for the LinearReLU modules
in static quantization mode. This includes torch.nn.qat.Linear,
torch.nn.intrinsic.LinearReLU, and torch.nn.intrinsic.qat.LinearReLU.
Future commits will add support for dynamic quantization and functional
LinearReLU.

Test Plan: python test/test_quantization.py TestQuantizeFx.test_linear_relu_lowering

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T107568666

[ghstack-poisoned]

jerryzh168 commented on test/quantization/fx/test_equalize_fx.py:
```diff
@@ -643,14 +644,14 @@ def test_input_weight_equalization_graphs(self):
             ns.call_function(torch.mul),
             ns.call_function(torch.quantize_per_tensor),
-            ns.call_module(nniq.LinearReLU),
+            ns.call_module(nni.modules.fused.LinearReLU),
```
this is the same as: nni.LinearReLU

jerryzh168 replied:
actually these should be lowered I feel, are we using reference models here?

andrewor14 replied:
As discussed offline, these changes will be reverted

jerryzh168 commented on test/quantization/fx/test_equalize_fx.py:
```diff
@@ -643,14 +644,14 @@ def test_input_weight_equalization_graphs(self):
             ns.call_function(torch.mul),
             ns.call_function(torch.quantize_per_tensor),
-            ns.call_module(nniq.LinearReLU),
+            ns.call_module(nni.modules.fused.LinearReLU),
```
same here and for occurrences

andrewor14 committed:
Update on "Add lowering path for LinearReLU module"


Summary: This commit adds a lowering path for the LinearReLU modules
in static quantization mode. This includes torch.nn.qat.Linear,
torch.nn.intrinsic.LinearReLU, and torch.nn.intrinsic.qat.LinearReLU.
Future commits will add support for dynamic quantization and functional
LinearReLU.

Test Plan: python test/test_quantization.py TestQuantizeFx.test_linear_relu_lowering

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T107568666

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33694742).

andrewor14 committed:
Update on "Add lowering path for LinearReLU module"


Summary: This commit adds a lowering path for the LinearReLU modules
in static quantization mode. This includes torch.nn.qat.Linear,
torch.nn.intrinsic.LinearReLU, and torch.nn.intrinsic.qat.LinearReLU.
Future commits will add support for dynamic quantization and functional
LinearReLU.

Test Plan: python test/test_quantization.py TestQuantizeFx.test_linear_relu_lowering

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T107568666

Differential Revision: [D33694742](https://our.internmc.facebook.com/intern/diff/D33694742)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33694742).

andrewor14 committed:
Update on "Add lowering path for LinearReLU module"


Summary: This commit adds a lowering path for the LinearReLU modules
in static quantization mode. This includes torch.nn.qat.Linear,
torch.nn.intrinsic.LinearReLU, and torch.nn.intrinsic.qat.LinearReLU.
Future commits will add support for dynamic quantization and functional
LinearReLU.

Test Plan: python test/test_quantization.py TestQuantizeFx.test_linear_relu_lowering

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T107568666

Differential Revision: [D33694742](https://our.internmc.facebook.com/intern/diff/D33694742)

[ghstack-poisoned]

andrewor14 committed:
Update on "Add lowering path for LinearReLU module"


Summary: This commit adds a lowering path for the LinearReLU modules
in static quantization mode. This includes torch.nn.qat.Linear,
torch.nn.intrinsic.LinearReLU, and torch.nn.intrinsic.qat.LinearReLU.
Future commits will add support for dynamic quantization and functional
LinearReLU.

Test Plan: python test/test_quantization.py TestQuantizeFx.test_linear_relu_lowering

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T107568666

Differential Revision: [D33694742](https://our.internmc.facebook.com/intern/diff/D33694742)

[ghstack-poisoned]

andrewor14 committed:
Update on "Add lowering path for LinearReLU module"


Summary: This commit adds a lowering path for the LinearReLU modules
in static quantization mode. This includes torch.nn.qat.Linear,
torch.nn.intrinsic.LinearReLU, and torch.nn.intrinsic.qat.LinearReLU.
Future commits will add support for dynamic quantization and functional
LinearReLU.

Test Plan: python test/test_quantization.py TestQuantizeFx.test_linear_relu_lowering

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T107568666

Differential Revision: [D33694742](https://our.internmc.facebook.com/intern/diff/D33694742)

[ghstack-poisoned]

andrewor14 committed:
Update on "Add lowering path for LinearReLU module"


Summary: This commit adds a lowering path for the LinearReLU modules
in static quantization mode. This includes torch.nn.qat.Linear,
torch.nn.intrinsic.LinearReLU, and torch.nn.intrinsic.qat.LinearReLU.
Future commits will add support for dynamic quantization and functional
LinearReLU.

Test Plan: python test/test_quantization.py TestQuantizeFx.test_linear_relu_lowering

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T107568666

Differential Revision: [D33694742](https://our.internmc.facebook.com/intern/diff/D33694742)

[ghstack-poisoned]

andrewor14 committed:
Update on "Add lowering path for LinearReLU module"


Summary: This commit adds a lowering path for the LinearReLU modules
in static quantization mode. This includes torch.nn.qat.Linear,
torch.nn.intrinsic.LinearReLU, and torch.nn.intrinsic.qat.LinearReLU.
Future commits will add support for dynamic quantization and functional
LinearReLU.

Test Plan: python test/test_quantization.py TestQuantizeFx.test_linear_relu_lowering

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T107568666

Differential Revision: [D33694742](https://our.internmc.facebook.com/intern/diff/D33694742)

[ghstack-poisoned]

andrewor14 committed:
Update on "Add lowering path for LinearReLU module"


Summary: This commit adds a lowering path for the LinearReLU modules
in static quantization mode. This includes torch.nn.qat.Linear,
torch.nn.intrinsic.LinearReLU, and torch.nn.intrinsic.qat.LinearReLU.
Future commits will add support for dynamic quantization and functional
LinearReLU.

Test Plan: python test/test_quantization.py TestQuantizeFx.test_linear_relu_lowering

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T107568666

Differential Revision: [D33694742](https://our.internmc.facebook.com/intern/diff/D33694742)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33694742).

jerryzh168 commented on torch/ao/quantization/fx/quantization_patterns.py:
```diff
@@ -917,18 +917,23 @@ def convert(self,
             # will remove the else branch after we migrated all use cases
-            if is_reference or \
-               type(self.linear) in [torch.nn.Linear] and dtypes in [(torch.quint8, torch.qint8, None)]:
+            module_whitelist = [
```
nit: whiltelist --> allowlist

jerryzh168 commented on torch/nn/intrinsic/quantized/modules/linear_relu.py:
```diff
@@ -34,3 +34,7 @@ def _get_name(self):
+
+    @classmethod
+    def from_reference(cls, ref_linear_relu, output_scale, output_zero_point):
+        return super().from_reference(ref_linear_relu[0], output_scale, output_zero_point)
```
would this return a nn.quantized.Linear module?

jerryzh168 replied:
oh nvm, we are using `cls` in `from_reference`

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -122,7 +141,7 @@ def _lower_to_native_backend(model: QuantizedGraphModule) -> QuantizedGraphModul
     operator signature so they can be lowered with the same function
     """
-    for ref_class in LOWER_MODULE_MAP.keys():
+    for ref_class in list(LOWER_MODULE_MAP.keys()) + list(LOWER_FUSED_MODULE_MAP):
```
do you need to call `.keys()` on LOWER_FUSED_MODULE_MAP as well?

andrewor14 replied:
Oh yes, `LOWER_FUSED_MODULE_MAP` used to be a set so I forgot to change this one too

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -122,7 +141,7 @@ def _lower_to_native_backend(model: QuantizedGraphModule) -> QuantizedGraphModul
     operator signature so they can be lowered with the same function
     """
-    for ref_class in LOWER_MODULE_MAP.keys():
+    for ref_class in list(LOWER_MODULE_MAP.keys()) + list(LOWER_FUSED_MODULE_MAP):
```
also, I'm wondering if we can just move this for loop to `_lower_weighted_ref_module` itself?

andrewor14 committed:
Update on "Add lowering path for LinearReLU module"


Summary: This commit adds a lowering path for the LinearReLU modules
in static quantization mode. This includes torch.nn.qat.Linear,
torch.nn.intrinsic.LinearReLU, and torch.nn.intrinsic.qat.LinearReLU.
Future commits will add support for dynamic quantization and functional
LinearReLU.

Test Plan: python test/test_quantization.py TestQuantizeFx.test_linear_relu_lowering

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T107568666

Differential Revision: [D33694742](https://our.internmc.facebook.com/intern/diff/D33694742)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33694742).

andrewor14 committed:
Update on "Add lowering path for LinearReLU module"


Summary: This commit adds a lowering path for the LinearReLU modules
in static quantization mode. This includes torch.nn.qat.Linear,
torch.nn.intrinsic.LinearReLU, and torch.nn.intrinsic.qat.LinearReLU.
Future commits will add support for dynamic quantization and functional
LinearReLU.

Test Plan: python test/test_quantization.py TestQuantizeFx.test_linear_relu_lowering

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T107568666

Differential Revision: [D33694742](https://our.internmc.facebook.com/intern/diff/D33694742)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33694742).

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -3269,24 +3269,70 @@ def __init__(self):
+        _test(qat=False)
+        _test(qat=True)
+
+    def test_linear_relu_lowering(self):
```
we do have some existing tests that already tests this path. maybe another thing we can do is to clean up those tests a bit:
https://github.com/pytorch/pytorch/blob/master/test/quantization/fx/test_quantize_fx.py#L3609

andrewor14 replied:
I see. I guess the main difference is this test explicitly compares the lowered and non-lowered models. I just modeled it after `test_linear_lowering`. I wonder if it makes sense to merge these two tests?

andrewor14 replied:
As discussed offline, we will try to merge this test with `test_linear_module`

andrewor14 committed:
Update on "Add lowering path for LinearReLU module"


Summary: This commit adds a lowering path for the LinearReLU modules
in static quantization mode. This includes torch.nn.qat.Linear,
torch.nn.intrinsic.LinearReLU, and torch.nn.intrinsic.qat.LinearReLU.
Future commits will add support for dynamic quantization and functional
LinearReLU.

Test Plan: python test/test_quantization.py TestQuantizeFx.test_linear_relu_lowering

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T107568666

Differential Revision: [D33694742](https://our.internmc.facebook.com/intern/diff/D33694742)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33694742).

andrewor14 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -3260,34 +3260,6 @@ def forward(self, x):
-        self.checkGraphModuleNodes(m, expected_node_occurrence=expected_node_occurrence)
-
-        # checking result match
-        self.assertEqual(out_ref, out)
```
This test is removed because its functionality is duplicated here:
https://github.com/pytorch/pytorch/blob/53c6917bd874952ae99577b3f705ae173f4ebac6/test/quantization/fx/test_quantize_fx.py#L3581

andrewor14 committed:
Update on "Add lowering path for LinearReLU module"


Summary: This commit adds a lowering path for the LinearReLU modules
in static quantization mode. This includes torch.nn.qat.Linear,
torch.nn.intrinsic.LinearReLU, and torch.nn.intrinsic.qat.LinearReLU.
Future commits will add support for dynamic quantization and functional
LinearReLU.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_linear_module

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T107568666

Differential Revision: [D33694742](https://our.internmc.facebook.com/intern/diff/D33694742)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33694742).

andrewor14 committed:
Update on "Add lowering path for LinearReLU module"


Summary: This commit adds a lowering path for the LinearReLU modules
in static quantization mode. This includes torch.nn.qat.Linear,
torch.nn.intrinsic.LinearReLU, and torch.nn.intrinsic.qat.LinearReLU.
Future commits will add support for dynamic quantization and functional
LinearReLU.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_linear_module

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Tasks: T107568666

Differential Revision: [D33694742](https://our.internmc.facebook.com/intern/diff/D33694742)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33694742).
